### ISSUE TITLE
[RFC][ResourceBundle] Loading "drivers"

### ISSUE THREAD

**Issue by aRn0D** [Sylius/Sylius#2622]
First, we need remove the `DatabaseDriverFactory` because it is not easily extendable. I think we should use a kind of chain of responsibility pattern because it allow us to configure several drivers (Doctrine and custom drivers). Every drivers must implement the same interface which ask to create the controller, repository and the manager as a service. Is the `DatabaseDriverInterface` is enough? I am sure because it ask to *load* the driver but it does not ask to *create* the right services.

Thus, what do we do ?

1 - Update the `DatabaseDriverInterface`, we should rename it in `ResourceDriverInterface`

```php
ResourceDriverInterface 
{
   // Create the controller definition
   public function createControllerDefinition($classe);

   // Create the object manager definition
   public function createManagerDefinition($classe);

   // Create the repository definition
   public function createRepositoryDefinition($classe);

   // Return the name of the name
   public function getSupportedDriver();
}
```

2 - Remove the `DatabaseDriverFactory`

3 - Declare every drivers as service (service definition file : driver.xml) with a tag `sylius_resource.driver_name`, *driver_name* is the name of the driver and it should match the name defined by `getSupportedDriver`.

```xml
        <!-- Resources/config/driver.xml -->
        <service id="id" class="%class%">
            <tag name="sylius_resource.custom_driver" />
        </service>
```

4 - Create a compiler pass `LoadResourceDriverPass`, for each resource it will get the right driver and create the right services
 
```php
namespace Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler;

class ObjectToIdentifierServicePass implements CompilerPassInterface
{
   public function process(ContainerBuilder $container)
   {
        $resources = container->getParameter('sylius.resource.config');

        foreach ($resources as $config) {
            // Get the driver
            $driver = $taggedServices = $container->findTaggedServiceIds('sylius_resource.driver');

            // Create the service that the resource bundle need to work
            $driver->addMethodCall('createControllerDefinition', array($config));
            $driver->addMethodCall('createManagerDefinition', array($config));
            $driver->addMethodCall('createRepositoryDefinition', array($config));
        }
    }
}
```

5 - We need to create `ResourceRepositoryInterface` `RecourceManagerInterface` which will be a generics interfaces for repositories and object managers. They will reduce the coupling to Doctrine and permit create custom driver easily.

@Sylius/core-team @stloyd What do you think ?

Related issue/PR #253, #2520

**Comment by amenophis** [comment]
Seems a good idea, 

About the part 3, i see something like this:
```xml
<!-- Resources/config/driver.xml -->
<service id="id" class="%class%">
    <tag name="sylius_resource.driver" alias="custom_driver" />
</service>
```

**Comment by stloyd** [comment]
@aRn0D Sounds good to me, and I agree with @amenophis about adding alias or something similar to tag declaration.

**Comment by aRn0D** [comment]
**Version 2**

First, we need remove the `DatabaseDriverFactory` because it is not easily extendable. I think we should use a kind of chain of responsibility pattern because it allow us to configure several drivers (Doctrine and custom drivers). Every drivers must implement the same interface which ask to create the controller, repository and the manager as a service. Is the `DatabaseDriverInterface` is enough? I am sure because it ask to *load* the driver but it does not ask to *create* the right services.

Thus, what do we do ?

1 - Update the `DatabaseDriverInterface`, we should rename it in `ResourceDriverInterface`

```php
ResourceDriverInterface 
{
   // Create the controller definition
   public function createControllerDefinition($classe);

   // Create the object manager definition
   public function createManagerDefinition($classe);

   // Create the repository definition
   public function createRepositoryDefinition($classe);

   // Return the name of the name
   public function getSupportedDriver();
}
```

2 - Remove the `DatabaseDriverFactory`

3 - Declare every drivers as service (service definition file : driver.xml) with a tag `sylius_resource.driver`, it should have a alias which will be the name of the driver and it should match the name defined by `getSupportedDriver`.

```xml
        <!-- Resources/config/driver.xml -->
        <service id="id" class="%class%">
            <tag name="sylius_resource.driver" alias="custom_driver" />
        </service>
```

4 - Create a compiler pass `LoadResourceDriverPass`, for each resource it will get the right driver and create the right services
 
```php
namespace Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler;

class ObjectToIdentifierServicePass implements CompilerPassInterface
{
   public function process(ContainerBuilder $container)
   {
        $resources = container->getParameter('sylius.resource.config');
        $drivers = $container->findTaggedServiceIds('sylius_resource.driver');

        foreach ($resources as $config) {
            $driver = /** Get the right driver for the current resource*/

            // Create the service that the resource bundle need to work
            $driver->addMethodCall('createControllerDefinition', array($config));
            $driver->addMethodCall('createManagerDefinition', array($config));
            $driver->addMethodCall('createRepositoryDefinition', array($config));
        }
    }
}
```

5 - We need to create `ResourceRepositoryInterface` `RecourceManagerInterface` which will be a generics interfaces for repositories and object managers. They will reduce the coupling to Doctrine and permit create custom driver easily.

**Comment by aRn0D** [comment]
@pjedrzejewski

**Comment by pjedrzejewski** [comment]
@aRn0D Sounds really cool, could we please just wait for v0.14.0? I am fighting with the translations now and I would like to get it fixed first and then we rework the resource, I totally agree!

**Comment by aRn0D** [comment]
@pjedrzejewski yes, sure it is why I just write rfc for now :)

**Comment by aRn0D** [comment]
@pjedrzejewski When do you plan to finish with the translations?

**Comment by Shine-neko** [comment]
@pjedrzejewski status ? :(

**Comment by pjedrzejewski** [comment]
Part of this implemented in #3721 but having custom drivers in container, which is not yet built is quite difficult or even impossible with current approach. We could have a "final" compiler pass etc. but that really complicates things. I think current approach is simplest. I will think about #253 which would partially solve the problem.
